### PR TITLE
fix: extend zx logs BI + generating real process errors (MAPCO-3950)

### PR DESCRIPTION
**Pull request by ronenkapelian**
Improve seed process logs interpretation + BI + errors

| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✔                                                                        |
| New feature     | ✔                                                                        |
| Breaking change | ✖                                                                        |
| Deprecations    | ✖                                                                        |
| Documentation   | ✖                                                                        |
| Tests added     | ✔                                                                        |
| Chore            | ✖                                                                       |

1. Use stream pipelines to handler cmd data instead of for await.
2. Add wrap that can able detect and handler process errors that was crashed the Node on beta version of seeder.
3. translate cli stdout into INFO and ERROR logs + evenets:
 a. Print only progress logs as info (by regex - "( /d tiles)".
 b. Catch seed errors by detecting logs (by regex " -  ERROR  - ") and failed as design.
4. test coverage unfortentaly was decrised :'(
:( 
becouse of zx mocking + process mocking complexity

### CODE REVIEW

File: src/common/interfaces.ts [hunk 1]
```diff
@@ -1,9 +1,14 @@
 /* eslint-disable @typescript-eslint/naming-convention */
-
 import { GeoJSON } from 'geojson';
-import { JsonObject } from 'swagger-ui-express';
 import { CacheType, SeedMode } from './enums';
 
+type JSONValue = string | number | boolean | JSONObject | JSONArray;
+interface JSONArray extends Array<JSONValue> {}
```
Comment (asafMasa) [inline src/common/interfaces.ts:6]: Lets talk about the need of this types in a more common library such as "https://github.com/MapColonies/types" or utils

File: src/mapproxyUtils/mapproxySeed.ts [hunk 2]
```diff
@@ -242,17 +242,66 @@ export class MapproxySeed {
...
+
       const cmd = $`mapproxy-seed ${flags}`;
+      // promise wrap to synchronized zx internal events with node run time event
+      await new Promise<void>((resolve, reject) => {
+        cmd.stdout
+          .on('data', (chunk: Buffer) => {
+            try {
+              this.seedProgressFunc(chunk);
+            } catch (error) {
+              // this section catching seeding errors and fail the task by kill child process of cmd
```
Comment (asafMasa) [inline src/mapproxyUtils/mapproxySeed.ts:254]: can't we log the error here? (with jobId, ...)

File: src/mapproxyUtils/mapproxySeed.ts [hunk 3]
```diff
@@ -242,17 +242,66 @@ export class MapproxySeed {
...
+              void cmd.kill('SIGHUP');
+              cmd.child.on('exit', () => {
+                reject(new Error((error as Error).message));
+              });
+            }
+          })
+          .on('end', () => {
+            this.logger.info(`Completed ${options.mode} task`);
```
Comment (asafMasa) [inline src/mapproxyUtils/mapproxySeed.ts:262]: add main job prams to all logs ...

File: src/mapproxyUtils/mapproxySeed.ts [hunk 4]
```diff
@@ -242,17 +242,66 @@ export class MapproxySeed {
...
+      await new Promise<void>((resolve, reject) => {
+        cmd.stdout
+          .on('data', (chunk: Buffer) => {
+            try {
+              this.seedProgressFunc(chunk);
+            } catch (error) {
+              // this section catching seeding errors and fail the task by kill child process of cmd
+              void cmd.kill('SIGHUP');
+              cmd.child.on('exit', () => {
+                reject(new Error((error as Error).message));
```
Comment (asafMasa) [inline src/mapproxyUtils/mapproxySeed.ts:257]: can't we log the error here? (with jobId, ...)

File: src/mapproxyUtils/mapproxySeed.ts [hunk 5]
```diff
@@ -242,17 +242,66 @@ export class MapproxySeed {
...
-      for await (const chunk of cmd.stdout) {
-        // todo - implement on next phase progressPercentage calculate logic
-        const str: string = (chunk as Buffer).toString('utf8');
-        this.logger.debug(str);
-      }
+      // promise wrap zx event on case of internal crash - so the node will catch up and throw exception to main loop
+      const exitCode = await cmd.exitCode;
+      await new Promise<void>((resolve, reject) => {
+        cmd.catch((error) => {
+          if ((error as ProcessOutput).exitCode !== 0) {
```
Comment (asafMasa) [inline src/mapproxyUtils/mapproxySeed.ts:271]: can't we log the error here (before the if)? (with jobId, ...)

File: src/mapproxyUtils/mapproxySeed.ts [hunk 6]
```diff
@@ -242,17 +242,66 @@ export class MapproxySeed {
...
+    const seedLogStr = chunk.toString('utf8');
+    this.logger.debug(seedLogStr); // print all mapproxy-seed stdout
+    if (seedLogStr.match(/- ERROR -/g)) {
+      // task will fail on case of seeding logic error (for example redis connection)
+      const errMsg = seedLogStr.split('- ERROR -')[1];
+      this.logger.error(errMsg);
+      throw new Error(errMsg);
+    }
+    if (seedLogStr.match(/\((\d)+ tiles\)/g)) {
+      this.logger.info(seedLogStr); // print only progress logs
```
Comment (asafMasa) [inline src/mapproxyUtils/mapproxySeed.ts:304]: how should we add job data to this logs?

**Review by asafMasa — CHANGES_REQUESTED** [review]
looking good, thanks! 🥇

File: src/mapproxyUtils/mapproxySeed.ts (hunk 2 above)
Comment (ronenkapelian) [reply src/mapproxyUtils/mapproxySeed.ts:254]: added

File: src/mapproxyUtils/mapproxySeed.ts (hunk 4 above)
Comment (ronenkapelian) [reply src/mapproxyUtils/mapproxySeed.ts:257]: already added on line 255

File: src/mapproxyUtils/mapproxySeed.ts (hunk 3 above)
Comment (ronenkapelian) [reply src/mapproxyUtils/mapproxySeed.ts:262]: added

File: src/mapproxyUtils/mapproxySeed.ts (hunk 5 above)
Comment (ronenkapelian) [reply src/mapproxyUtils/mapproxySeed.ts:271]: added

File: src/mapproxyUtils/mapproxySeed.ts (hunk 6 above)
Comment (ronenkapelian) [reply src/mapproxyUtils/mapproxySeed.ts:304]: it is TBD, i kept the method simplified with params, till we will design and understand the future changing

File: src/common/interfaces.ts (hunk 1 above)
Comment (asafMasa) [reply src/common/interfaces.ts:6]: open an issue

**Review by asafMasa — COMMENTED** [review]
👍

**Review by asafMasa — APPROVED** [review]

**Pull request merged**